README.md
# code-buster

A Clojure library designed to ... well, that part is up to you.

# Rules


The game is played on a map 16001 units wide and 9001 units high. 
The coordinate X=0, Y=0 is the top left pixel.

Each player controls a team of several busters. The teams start out at opposite corners of
the map, near their base. Ghosts are scattered across the map, and need to be trapped and 
brought back to a player's base. Each ghost inside the base or trapped by a buster is worth
a point for the owner. However, you may lose a point if one of your busters releases 
a ghost outside of your base.

**The map works as follows:**

* There are always 2 teams in play.
* At the start of the game, each player is given a team id. This indicates on which corner their base is positioned. The top left corner (X=0, Y=0) is for team 0. The bottom right corner ( X=16000, Y=9000) is for team 1.
* Fog makes it impossible to know the positions of ghosts and rival busters unless they are within 2200 units from one of your own busters.
* Each buster has a unique id. Each ghost has a unique id. Ghosts and busters with the same id are not correlated.

**Busters work as follows:**

* Every turn, a buster may do one of the following actions: `MOVE`, `BUST`, `RELEASE` or `STUN`
* `MOVE` followed by map coordinates will make the buster advance 800 units towards that point. The position will be rounded to the nearest integer.
* `BUST` followed by a ghost's id will cause the buster to suck that ghost into his ghost trap if he is within a range of 1760 units around the target ghost but not closer than 900 units. Trapped ghosts are no longer visible on the map.
* A buster may carry no more than 1 ghost at a time.
* `RELEASE` will cause the buster to drop the ghost he is carrying back onto the map. If this is done within 1600 units of a corner of the map on which a base is positioned, the ghost is removed from play and secures a point for the base's owner.
* `STUN` followed by a buster's id will make the buster shoot a bolt of lightning that will stun the target buster. A stunned buster cannot do anything for 10 turns. The buster's stun ability must recharge for 20 turns before it can be used again. Being stunned or attempting to stun will cause a buster to release any ghost he may be carrying.
* A buster may only stun opponents within a 1760 unit radius.

**Ghosts work as follows:**

* Ghosts remain stationary unless busters are within 2200 units of them and they are not in the process of being trapped. In this case, they will move 400 units away from the mean position of the near busters.
* If several busters attempt to trap a ghost with 0 stamina, the team with the most busters will take priority. Within that team, the closest buster will take the ghost. If both teams have an equal number of busters attempting to trap the ghost, it will not be trapped on that turn.
* A ghost being carried by a buster will escape if that buster attempts to trap a different ghost.
* Stamina does not regenerate.
* The game ends once all ghosts have been captured or after the time limit of 400 turns.

The game state of every turn is given to you as a list of entities, each with an id, position, type, state and value.

The type will be:

* 0 for a buster from team 0.
* 1 for a buster from team 1.
* -1 for a ghost.

The state will be:

* For busters:
  * 0: idle or moving buster.
  * 1: buster carrying a ghost.
  * 2: stunned buster.
  * 3: buster in the process of trapping a ghost.
* For ghosts, it is equal to the ghost's current stamina.

The value will be:

* For busters:
  * If the buster is carrying a ghost, that ghost's id.
  * If the buster is stunned the number of turns until he can move again.
* For ghosts, it is 0 unless several busters tied in trying to trap it, in which cas it is equal to the amount of busters that attempted to trap it on that turn.
 
**Victory Conditions**

Have captured more ghosts than the rival team at the end of the game.
 
**Lose Conditions**

* Your program provides unrecognized output.
* Your program times out.
* You have less ghosts than the opponent at the end of the game.
 
## Note

The program must first read within an infinite loop, read the contextual data from the standard input and provide to the standard output the desired instructions.
 	Game Input

**Initialization input**

* Line 1: an integer bustersPerPlayer for the amount of busters each team controls.
* Line 2: an integer ghostCount for the amount of ghosts available on the map.
* Line 3: an integer myTeamId for the team you are playing as.

**Input for one game turn**

* Line 1: an integer entities for the amount of entities on the map that are visible to you.
* Next entities lines: 6 space separated integers entityId, x, y, entityType, state & value. 

Represents a buster or a ghost.

**Output for one game turn**

One line for each buster: one of the following:

* `MOVE` followed by two integers x and y
* `BUST` followed by one integer ghostId
* `RELEASE`
* `STUN` followed by one integer busterId

You may append text to your instructions, it will be displayed in the viewer.

Constraints

* 2 ≤ bustersPerPlayer ≤ 5
* 8 ≤ ghostCount ≤ 28

Response time per turn ≤ 100ms

What is in store for me in the higher leagues ?
The extra rule available in higher leagues is:
Ghosts with endurance, who take several turns to successfully trap.

## Developers

    (require '[clojure.test :refer [run-tests]])
    (require 'Player-test)
    (run-tests 'Player-test)
    (Player-test/do-think-case-3)
    
    
* [Using Potential Fields in a Real-time Strategy Game Scenario](http://aigamedev.com/open/tutorials/potential-fields/)
* 

## License

Copyright © 2016 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2016-06-26
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2016-06-26
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/code-buster/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/code-buster/compare/0.1.0...0.1.1


src/Drawings.clj
(ns Drawings
  (:require [quil.core :as q]
            [Player :as p]))
;;------------------------------------------------------------------------------------------

(def pos (atom {:x (/ p/WIDTH 2)
                :y (/ p/HEIGHT 2)}))
(def v (atom {:vx 150 :vy 100}))
(def g (atom (p/generate-grid)))

;;------------------------------------------------------------------------------------------

(defn- s [x]
  "basic scaling..."
  (int (/ x 20)))

(defn setup []
  (q/frame-rate 10)                                         ;; Set framerate to 1 FPS
  (q/background 255))                                       ;; Set the background colour to

;; a nice shade of grey.
(defn draw-p []
  (q/stroke (q/random 255))                                 ;; Set the stroke colour to a random grey
  (q/stroke-weight (q/random 10))                           ;; Set the stroke thickness randomly
  (q/fill (q/random 255))                                   ;; Set the fill colour to a random grey

  (let [diam (q/random 100)                                 ;; Set the diameter to a value between 0 and 100
        x (q/random (q/width))                              ;; Set the x coord randomly within the sketch
        y (q/random (q/height))]                            ;; Set the y coord randomly within the sketch
    (q/ellipse x y diam diam)))                             ;; Draw a circle at x y with the correct diameter

(defn draw []
  (q/clear)
  (let [ex (+ (:x @pos) (:vx @v))
        ey (+ (:y @pos) (:vy @v))
        nvx (if (or (< ex 0) (< p/WIDTH ex)) (- (:vx @v)) (:vx @v))
        nvy (if (or (< ey 0) (< p/HEIGHT ey)) (- (:vy @v)) (:vy @v))

        e1 {:x ex :y ey}
        es [e1]
        grid (-> @g
                 (p/tick-cells)
                 (p/mark-visited-cells e1)
                 (p/compute-zones-of-interest))
        diam (p/grid-cell-diameter grid)]

    (reset! pos {:x ex :y ey})
    (reset! v {:vx nvx :vy nvy})
    (reset! g grid)
    (doseq [z (p/grid-cells grid)]
      (q/stroke 34)
      (q/stroke-weight 2)
      ;(q/no-fill)
      (q/fill (+ 160 (* 4 (max 0 (:visited z))))
              (- 160 (* 8 (max 0 (:visited z))))
              (- 160 (* 8 (max 0 (:visited z))))
              255)
      (q/ellipse (s (:x z))
                 (s (:y z))
                 (s diam)
                 (s diam)))
    (q/fill 200 100)
    (doseq [e es]
      (q/ellipse (s (:x e))
                 (s (:y e))
                 (s p/FOG_DIAMETER)
                 (s p/FOG_DIAMETER)))))

(q/defsketch example                                        ;; Define a new sketch named example
             :title "Oh so many grey circles"               ;; Set the title of the sketch
             :settings #(q/smooth 2)                        ;; Turn on anti-aliasing
             :setup setup                                   ;; Specify the setup fn
             :draw draw                                     ;; Specify the draw fn
             :size [(s p/WIDTH) (s p/HEIGHT)])              ;; You struggle to beat the golden ratio



src/Player.clj
(ns Player
  (:gen-class)
  (:import (clojure.lang LineNumberingPushbackReader)
           (java.io StringReader)))

;
; TODO rethink exploration... follow enemies?
; TODO keep track of seen enemies to hunt them
; TODO try roles: Explorer, Hunter, ...
;
;


;---------------------------------------------------------------
;                _   _ _ _ _   _
;          _   _| |_(_) (_) |_(_) ___  ___
;    _____| | | | __| | | | __| |/ _ \/ __|
;   |_____| |_| | |_| | | | |_| |  __/\__ \
;          \__,_|\__|_|_|_|\__|_|\___||___/
;---------------------------------------------------------------

(defn as-stream [s]
  (-> (StringReader. s) LineNumberingPushbackReader.))

(defn debug [& args]
  (binding [*out* *err*]
    (println args)))

(defn sqr [x]
  (* x x))

(defn- d-sqr [c1 c2]
  (sqr (- c1 c2)))

(defn d2 [e1 e2]
  (+ (d-sqr (:x e1) (:x e2))
     (d-sqr (:y e1) (:y e2))))

(defn distance [e1 e2]
  (int (Math/sqrt (d2 e1 e2))))

(defn abs [x]
  (if (neg? x) (- x)
               x))

;---------------------------------------------------------------
;       _      _                      _   _     _       _
;      | | ___| |_   _ __ ___   ___  | |_| |__ (_)_ __ | | __
; _____| |/ _ \ __| | '_ ` _ \ / _ \ | __| '_ \| | '_ \| |/ /
;|_____| |  __/ |_  | | | | | |  __/ | |_| | | | | | | |   <
;      |_|\___|\__| |_| |_| |_|\___|  \__|_| |_|_|_| |_|_|\_\
;
;---------------------------------------------------------------
(def WIDTH 16001)
(def HEIGHT 9001)
(def MIN_DIST_BUST 900)
(def MAX_DIST_BUST 1760)
(def MIN_D2_BUST (sqr MIN_DIST_BUST))
(def MAX_D2_BUST (sqr MAX_DIST_BUST))

(def MAX_DIST_RELEASE 1600)
(def MAX_D2_RELEASE (sqr MAX_DIST_RELEASE))

(def MAX_DIST_MOVE 800)

(def MAX_DIST_STUN 1760)
(def MAX_D2_STUN (sqr MAX_DIST_STUN))
(def MIN_DIST_STUN 0)
(def MIN_D2_STUN (sqr MIN_DIST_STUN))

(def FOG_RADIUS 2200)
(def FOG_RADIUS2 (sqr FOG_RADIUS))
(def FOG_DIAMETER (* 2 FOG_RADIUS))
(def HOME1 {:x 0 :y 0})
(def HOME2 {:x 16000 :y 9000})
(def STUN_RELOAD 20)
(def GHOST_MOVE_AWAY 400)

(def VISITED_ALIVE_DURATION 50)

(def STATE_IDLE 0)
(def STATE_CARRYING 1)
(def STATE_STUNNED 2)
(def STATE_TRAPPING 3)

(defn closest [buster positionables]
  (first (sort-by #(d2 buster %) positionables)))

(defn farest [buster positionables]
  (last (sort-by #(d2 buster %) positionables)))

(defn closest-to-bust [buster ghosts]
  (first (sort-by #(abs (- MIN_D2_BUST (d2 buster %))) ghosts)))


;---------------------------------------------------------------
; GRID
;---------------------------------------------------------------

(defn grid-cell-diameter [grid]
  (get grid :diameter))

(defn grid-cells [grid]
  (vals (get grid :cells)))

(defn generate-grid []
  (let [sz (* 0.1 FOG_DIAMETER)
        decal (/ sz 2)
        cells (for [x (range 0 (dec (/ WIDTH sz)))
                    y (range 0 (dec (/ HEIGHT sz)))]
                [[x y] {:x       (+ decal (* x sz))
                        :y       (+ decal (* y sz))
                        :visited 0}])
        cells (into {} cells)]
    {:diameter sz
     :cells    cells}))

(defn tick-cells [grid]
  (let [cells (:cells grid)
        newCells (reduce (fn [acc [k cell]]
                           (assoc acc k (update-in cell [:visited] dec))) {} cells)]
    (assoc grid :cells newCells)))

(defn mark-visited-cells [grid buster]
  (let [cells (:cells grid)
        diam (:diameter grid)
        limit (sqr (- FOG_RADIUS (/ diam 6)))
        newCells (reduce (fn [acc [k cell]]
                           (if (< (d2 buster cell) limit)
                             (assoc acc k (assoc cell :visited VISITED_ALIVE_DURATION))
                             (assoc acc k cell))) {} cells)]
    (assoc grid :cells newCells)))

(defn compute-zones-of-interest [grid]
  grid)

;---------------------------------------------------------------

(defn generate-zones-0 []
  (let [sz (* 0.75 FOG_DIAMETER)]
    (for [x (range 0 (/ WIDTH sz))
          y (range 0 (/ HEIGHT sz))]
      {:x (+ (/ FOG_RADIUS 2) (* x sz))
       :y (+ (/ FOG_RADIUS 2) (* y sz))})))


(defn generate-zones [myTeam]
  (let [zones [{:order 1, :x 1100.0, :y 1100.0, :weight (if (= 0 myTeam) 10 80)}
               {:order 12, :x 1100.0, :y 4400.0, :weight 100}
               {:order 11, :x 1100.0, :y 7700.0, :weight 10}
               {:order 2, :x 4400.0, :y 1100.0, :weight 10}
               {:order 13, :x 4400.0, :y 4400.0, :weight 10}
               {:order 10, :x 4400.0, :y 7700.0, :weight 10}
               {:order 3, :x 7700.0, :y 1100.0, :weight (if (= 0 myTeam) 50 20)}
               {:order 14, :x 7700.0, :y 4400.0, :weight 10}
               {:order 9, :x 7700.0, :y 7700.0, :weight (if (= 0 myTeam) 20 50)}
               {:order 4, :x 11000.0, :y 1100.0, :weight 10}
               {:order 15, :x 11000.0, :y 4400.0, :weight 10}
               {:order 8, :x 11000.0, :y 7700.0, :weight 10}
               {:order 5, :x 14300.0, :y 1100.0, :weight 100}
               {:order 6, :x 14300.0, :y 4400.0, :weight 10}
               {:order 7, :x 14300.0, :y 7700.0, :weight (if (= 0 myTeam) 80 10)}]]
    (if (= 0 myTeam)
      zones
      (map (fn [z] (let [w (:order z)]
                     (assoc z :order (- 16 w)))) zones))))

(defn generate-unexplored-zones [myTeam]
  [{:order (if (= 0 myTeam) 1 2), :x 1100.0, :y 1100.0}
   {:order (if (= 0 myTeam) 4 3), :x 1100.0, :y 7700.0}
   {:order (if (= 0 myTeam) 3 4), :x 14300.0, :y 1100.0}
   {:order (if (= 0 myTeam) 2 1), :x 14300.0, :y 7700.0}])

;---------------------------------------------------------------

(defn my-home [game]
  (if (= 0 (:myTeamId game))
    HOME1
    HOME2))

(defn init-ia-data [game]
  {:unvisited-zones  (generate-zones (:myTeamId game))
   :unexplored-zones (generate-unexplored-zones (:myTeamId game))
   :ghosts           []
   :busters          (reduce (fn [acc id] (assoc acc id {}))
                             {}
                             (range (:bustersPerPlayer game)))})


(defn my-busters [game entities]
  (let [myTeamId (:myTeamId game)]
    (filter #(= (:entityType %) myTeamId) entities)))

(defn other-busters [game entities]
  (let [myTeamId (:myTeamId game)]
    (filter #(not (or (= (:entityType %) myTeamId)
                      (= (:entityType %) -1))) entities)))

(defn ghosts [entities]
  (filter #(= (:entityType %) -1) entities))

(defn can-bust? [e1 e2]
  (let [ds (d2 e1 e2)]
    (and (<= ds MAX_D2_BUST)
         (<= MIN_D2_BUST ds))))

(defn stamina-of [ghost]
  (:state ghost))

(defn should-bust? [ghost]
  (let [nb-total-busters (:value ghost)
        nb-busters (:nb-busters ghost)
        nb-busters (if (nil? nb-busters) 0 nb-busters)
        stamina (stamina-of ghost)
        r (or
            (< nb-busters 1)
            (and (<= nb-busters nb-total-busters) (= 0 stamina))
            (< (* 2 nb-busters) stamina))]
    (debug "should-bust?" ghost)
    (debug "            " r "(" nb-busters " vs " stamina ")")
    r))

(defn must-release? [buster]
  (= STATE_CARRYING (:state buster)))

(defn stunned? [buster]
  (= STATE_STUNNED (:state buster)))

(defn trapping? [buster]
  (= STATE_TRAPPING (:state buster)))

(defn carrying? [buster]
  (= STATE_CARRYING (:state buster)))

(defn can-release? [game buster]
  (<= (d2 (my-home game) buster) MAX_D2_RELEASE))



(defn number-of-round-to-reach [e1 e2]
  (let [d (distance e1 e2)
        nb-round (Math/ceil (/ d MIN_DIST_BUST))]
    (int nb-round)))

(defn calculate-reachability [buster ghosts]
  (map #(assoc % :interest (max (stamina-of %)
                                (number-of-round-to-reach buster %))) ghosts))

(defn sort-by-reachability [buster ghosts]
  (let [gs (calculate-reachability buster ghosts)
        ]
    (debug "interests")
    (doseq [g gs]
      (debug "   stamina: " (stamina-of g) " i:" g))
    (sort-by :interest gs)))

(defn nb-round-before-reload [ia-data buster]
  (let [round (:round ia-data)
        stun-round (get-in ia-data [:busters (:entityId buster) :stun-round])]
    (if (nil? stun-round)
      0
      (max 0 (- (+ stun-round STUN_RELOAD) round)))))

(defn can-stun? [ia-data buster]
  (= 0 (nb-round-before-reload ia-data buster)))

(defn can-stun-target? [e1 e2]
  (if (and e1 e2)
    (let [d2 (d2 e1 e2)]
      (and (< d2 MAX_D2_STUN)
           (<= MIN_D2_STUN d2)))
    false))

(defn action-bust [ia-data buster e]
  {:type    :bust
   :ghostId (:entityId e)
   :msg     (str "Zzzappp! " (nb-round-before-reload ia-data buster))})

(defn action-release [ia-data buster]
  {:type :release
   :msg  (str "psschiitt! " (nb-round-before-reload ia-data buster))})

(defn action-move
  ([ia-data buster e ^String msg]
   (action-move ia-data buster (:x e) (:y e) msg))
  ([ia-data buster x y ^String msg]
   {:type :move
    :x    (int x)
    :y    (int y)
    :msg  (str msg " " (nb-round-before-reload ia-data buster))}))

(defn action-stun [ia-data buster target]
  {:type   :stun
   :target (:entityId target)
   :msg    (str "take that! " (nb-round-before-reload ia-data buster))})


(defn action-stun-other [ia-data buster targets]
  (let [others (:round-other-busters ia-data)
        target (first targets)]                             ; TODO: keep the one my other busters could not stun
    [(-> ia-data
         (update-in [:busters (:entityId buster)] assoc :stun-round (:round ia-data))
         (assoc :round-other-busters (remove #{target} others)))
     (action-stun ia-data buster target)]))

(defn action-move-random [ia-data buster]
  (let [seen-ghosts (:ghosts ia-data)
        target-ghost (first (sort-by-reachability buster seen-ghosts))]
    (if (seq target-ghost)
      [ia-data
       (action-move ia-data buster target-ghost "previously seen!")]

      (let [zones (:unvisited-zones ia-data)
            grouped (group-by :weight zones)
            unvisited (if (empty? grouped)
                        []
                        (get grouped (apply max (keys grouped))))
            target-zone (get-in ia-data [:busters (:entityId buster) :target-zone])
            preferred-zone (if (or (nil? target-zone)
                                   (not-any? #{target-zone} unvisited))
                             (farest buster unvisited)
                             target-zone)]
        (debug "#:preferred unvisited zone '" preferred-zone "' target-zone: " target-zone)
        (if preferred-zone
          [(-> ia-data
               (assoc :unvisited-zones (remove #(= % preferred-zone) zones))
               (update-in [:busters (:entityId buster)] assoc :target-zone preferred-zone))
           (action-move ia-data buster preferred-zone "let's explore!")]
          (let [random (rand-int 360)
                theta (/ (* Math/PI random) 180)
                x (+ (:x buster) (* MAX_DIST_MOVE (Math/cos theta)))
                y (+ (:y buster) (* MAX_DIST_MOVE (Math/sin theta)))]
            [ia-data (action-move ia-data buster x y "random move")]))))))

(defn action-move-away [ia-data buster dst]
  (let [deltaX (- (:x dst) (:x buster))
        deltaY (- (:y dst) (:y buster))
        dist (Math/sqrt (+ (sqr deltaX) (sqr deltaY)))
        deltaDist (- (+ dist GHOST_MOVE_AWAY) (/ (+ MAX_DIST_BUST MIN_DIST_BUST) 2))
        ux (/ deltaX dist)
        uy (/ deltaY dist)
        newX (+ (:x buster) (* deltaDist ux))
        newY (+ (:y buster) (* deltaDist uy))]
    (action-move ia-data buster newX newY (str "aside"))))

(defn action-move-to-release [ia-data buster game]
  (action-move ia-data buster (my-home game) "recycling!")
  ;(let [dh1 (d2 HOME1 buster)
  ;      dh2 (d2 HOME2 buster)]
  ;  (if (< dh1 dh2)
  ;    (action-move HOME1 "let's trash this")
  ;    (action-move HOME2 "let's trash this")))
  )

(defn action-stunned [ia-data buster]
  (action-move ia-data buster 1 1 "stunned!"))

(defn one-or-inc [x]
  (if (nil? x) 1 (inc x)))

(defn do-buster-think-0 [game ia-data buster]
  (let [ghosts (:round-ghosts ia-data)
        ghosts-bustable (filter #(and (can-bust? buster %)
                                      (should-bust? %)) ghosts)
        closest-ghost (closest buster ghosts)

        other-busters (:round-other-busters ia-data)
        stunnable-busters (filter #(and (not (stunned? %))
                                        (can-stun-target? buster %)) other-busters)
        high-priority-targets (filter #(or (trapping? %)
                                           (carrying? %)) stunnable-busters)

        canStun (can-stun? ia-data buster)

        _ (debug "---me:" buster)
        _ (debug "   canStun..........:" canStun)
        _ (debug "   VIP..............:" high-priority-targets)
        _ (debug "   ghost bustable...:" ghosts-bustable)]
    (cond (stunned? buster)
          [ia-data (action-stunned ia-data buster)]

          (and canStun
               (seq high-priority-targets))
          (action-stun-other ia-data buster high-priority-targets)

          (must-release? buster)
          (if (can-release? game buster)
            [ia-data (action-release ia-data buster)]
            (if (and canStun
                     (seq stunnable-busters))
              (action-stun-other ia-data buster stunnable-busters)
              [ia-data (action-move-to-release ia-data buster game)])) ; TODO move away from visible enemies


          (and (seq ghosts))
          (if (seq ghosts-bustable)
            (let [ghost (first (sort-by-reachability buster ghosts-bustable))
                  next-ia (-> ia-data
                              (assoc :round-ghosts
                                     (map (fn [b] (if (= b ghost)
                                                    (update-in b [:nb-busters] one-or-inc)
                                                    b)) ghosts)))]
              [next-ia (action-bust ia-data buster ghost)])
            [ia-data (action-move-away ia-data buster (closest-to-bust buster ghosts))])

          (and canStun
               (seq stunnable-busters))
          (action-stun-other ia-data buster stunnable-busters)


          (and closest-ghost)
          [(-> ia-data
               (assoc :ghosts (-> (remove #(= % closest-ghost) ghosts)
                                  (conj (assoc closest-ghost :buster (:entityId buster)))))
               (update-in [:busters (:entityId buster)] dissoc :target-zone))
           (action-move ia-data buster closest-ghost "a ghost!")]

          :else
          (action-move-random ia-data buster))))

(defn do-explore [game ia-data buster]
  (let [other-busters (:round-other-busters ia-data)
        stunnable-busters (filter #(and (not (stunned? %))
                                        (can-stun-target? % buster)) other-busters)
        canStun (can-stun? ia-data buster)
        ; ---

        zones (:unexplored-zones ia-data)
        ordered (sort-by :order zones)
        target-zone (get-in ia-data [:busters (:entityId buster) :target-zone])
        target-zone (if (or (nil? target-zone)
                            (< (d2 buster target-zone) 50))
                      (first ordered)
                      target-zone)
        _ (debug "---me:" buster)
        _ (debug "   unexplored" zones)
        _ (debug "   target" target-zone ", d=" (if target-zone (distance target-zone buster)))
        ]
    (cond
      (and canStun
           (seq stunnable-busters))
      (action-stun-other ia-data buster stunnable-busters)

      target-zone
      [(-> ia-data
           (assoc :unexplored-zones (remove #(= % target-zone) zones))
           (update-in [:busters (:entityId buster)] assoc :target-zone target-zone))
       (action-move ia-data buster target-zone "exploring!")]

      :else
      [(-> ia-data
           (update-in [:busters (:entityId buster)] dissoc :target-zone))
       nil])))

(defn do-buster-think [game ia-data buster]
  (let [role (get-in ia-data [:roles (:entityId buster)])
        [nia-data computed] (cond
                              (= :explorer role)
                              (do-explore game ia-data buster)
                              :else
                              [ia-data nil])]
    (if computed
      [nia-data computed]
      (do-buster-think-0 game nia-data buster))))



(defn update-ia-data [ia-data game entities]
  (let [my-busters (my-busters game entities)
        other-busters (other-busters game entities)
        ghosts (ghosts entities)
        ;
        unvisited (:unvisited-zones ia-data)
        visited-zones (reduce (fn [zs buster]
                                (into zs (filter #(< (d2 buster %) 50) unvisited)))
                              #{} my-busters)
        ;
        seen-ghosts (:ghosts ia-data)
        ; --- remove ghosts that should currently be seen, in case there were zapped
        ghosts-in-fog (reduce (fn [gs buster]
                                (into gs (filter #(< (d2 buster %) FOG_RADIUS2) seen-ghosts)))
                              #{} my-busters)]
    (-> ia-data
        (assoc :ghosts (into ghosts (remove ghosts-in-fog seen-ghosts)))
        (assoc :unvisited-zones (remove visited-zones unvisited))
        (assoc :round-ghosts ghosts)
        (assoc :round-busters my-busters)
        (assoc :round-other-busters other-busters))))

(defn do-think [game round-ia-data]
  (let [my-busters (:round-busters round-ia-data)
        [ia-data actions] (reduce (fn [[ia-data actions] buster]
                                    (let [x (do-buster-think game ia-data buster)
                                          [new-data action] x]
                                      [new-data (conj actions action)]))
                                  [round-ia-data []] my-busters)
        next-ia-data (-> round-ia-data
                         (assoc :busters (:busters ia-data))
                         (assoc :unexplored-zones (:unexplored-zones ia-data)))]
    [next-ia-data actions]))

;---------------------------------------------------------------
;             _
;            (_) ___
;       _____| |/ _ \
;      |_____| | (_) |
;            |_|\___/
;---------------------------------------------------------------

(defn read-initialization-input [in]
  {:bustersPerPlayer (read in)
   :ghostCount       (read in)
   :myTeamId         (read in)})

; Send your busters out into the fog to trap ghosts and bring them home!
(defn read-game-turn [game in]
  ; nb-entities: the number of busters and ghosts visible to you
  (let [nb-entities (read in)]
    (loop [i nb-entities
           entities []]
      (if (> i 0)
        (let [entity {:entityId   (read in)                 ; entityId: buster id or ghost id
                      :x          (read in)                 ; x, y: position of this buster / ghost
                      :y          (read in)
                      :entityType (read in)                 ; entityType: the team id if it is a buster, -1 if it is a ghost.
                      :state      (read in)                 ; state: For busters: 0=idle, 1=carrying a ghost.
                      :value      (read in)}]               ; value: For busters: Ghost id being carried. For ghosts: number of busters attempting to trap this ghost.
          (recur (dec i) (conj entities entity)))
        entities))))

(defn display-action [action]
  (cond
    (= :release (:type action)) (println "RELEASE" (:msg action))
    (= :move (:type action)) (println "MOVE" (int (:x action)) (int (:y action)) (:msg action))
    (= :bust (:type action)) (println "BUST" (:ghostId action) (:msg action))
    (= :stun (:type action)) (println "STUN" (:target action) (:msg action))))

(defn display-actions [actions]
  (doseq [action actions]
    (display-action action)))

(def EXPLORER true)

(defn affect-roles [game ia-data]
  (assoc ia-data
    :roles
    (cond-> {}
            EXPLORER (assoc 0 :explorer))))

(defn -main [& args]
  (let [game (read-initialization-input *in*)]
    (debug "game/input" game)
    ; bustersPerPlayer: the amount of busters you control
    ; ghostCount: the amount of ghosts on the map
    ; myTeamId: if this is 0, your base is on the top left of the map, if it is one, on the bottom right
    (loop [round 1
           ia-data (affect-roles game (init-ia-data game))]
      (debug "Round #" round)
      (let [entities (read-game-turn game *in*)
            _ (debug "game/entities" entities)
            _ (debug "game/ia-data" ia-data)
            round-ia-data (-> ia-data
                              (assoc :round round)
                              (update-ia-data game entities))
            [new-ia-data actions] (do-think game round-ia-data)]
        (debug "actions" actions)
        (display-actions actions)
        (recur (inc round)
               (dissoc new-ia-data
                       :round-ghosts
                       :round-busters
                       :round-other-busters))))))
